### PR TITLE
NIFI-2452

### PR DESCRIPTION
**Pull request by markap14**
Ensure that we keep track of how many references we have to each lucene searcher and only close the underlying index reader if there are no references to the searcher.  Also updated to prefer newer provenance events over older provenance events, and calculate FlowFile lineage based on an event id instead of a FlowFile UUID, as it's much more efficient

### CODE REVIEW

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/AbstractHeartbeatMonitor.java [hunk 1]
```diff
@@ -203,16 +203,24 @@ private void processHeartbeat(final NodeHeartbeat heartbeat) {
             final DisconnectionCode disconnectionCode = connectionStatus.getDisconnectCode();
 
             // Determine whether or not the node should be allowed to be in the cluster still, depending on its reason for disconnection.
-            if (disconnectionCode == DisconnectionCode.LACK_OF_HEARTBEAT || disconnectionCode == DisconnectionCode.UNABLE_TO_COMMUNICATE) {
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster/src/main/java/org/apache/nifi/cluster/coordination/heartbeat/AbstractHeartbeatMonitor.java]: was this heartbeat stuff supposed to be in this PR?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java [hunk 2]
```diff
@@ -16,7 +16,38 @@
  */
 package org.apache.nifi.controller;
 
-import com.sun.jersey.api.client.ClientHandlerException;
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java]: this set of changes also appears to not belong in this PR?

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/PersistentProvenanceRepository.java [hunk 3]
```diff
@@ -2166,6 +2171,28 @@ private Lineage computeLineage(final Collection<String> flowFileUuids, final NiF
...
+        }
+
+        if (event == null) {
+            final AsyncLineageSubmission result = new AsyncLineageSubmission(LineageComputationType.FLOWFILE_LINEAGE, eventId, Collections.<String> emptySet(), 1, user.getIdentity());
+            result.getResult().setError("Could not find Provenance Event with ID " + eventId);
+            lineageSubmissionMap.put(result.getLineageIdentifier(), result);
+            return result;
+        }
+
+        return submitLineageComputation(Collections.singleton(event.getFlowFileUuid()), user, LineageComputationType.FLOWFILE_LINEAGE, eventId, event.getLineageStartDate(), Long.MAX_VALUE);
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/PersistentProvenanceRepository.java]: so first we search on the eventId then we search on the flowfile itself.  Can you describe why/what is happening there?  Why isn't the first query sufficient?

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/test/java/org/apache/nifi/provenance/TestPersistentProvenanceRepository.java [hunk 4]
```diff
@@ -481,6 +485,165 @@ public void testIndexAndCompressOnRolloverAndSubsequentSearch() throws IOExcepti
         assertTrue(newRecordSet.getMatchingEvents().isEmpty());
     }
 
+    // TODO: Switch to 10,000.
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/test/java/org/apache/nifi/provenance/TestPersistentProvenanceRepository.java]: TODO Reminder?  Do you still want to do?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java (hunk 2 above)
Comment (markap14) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/FlowController.java]: @joewitt this change was due to a bug that I ran into when testing this PR. The bug was preventing me from moving forward, and was a rather trivial fix, so I just included it in the PR. I could have created a separate JIRA and PR but since it would have blocked this PR it seemed more prudent to just roll it in.

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/PersistentProvenanceRepository.java (hunk 3 above)
Comment (markap14) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/PersistentProvenanceRepository.java]: When we obtain the event by ID above, we are not querying anything - we are simply looking up the event directly by ID. We then formulate a query based on the UUID of the FlowFile. This is different than passing in the UUID of the FlowFile directly, though, because when we look up the Event directly, we also have the Lineage Start Date. Knowing that Lineage Start Date can dramatically reduce the amount of searching required if the start date is fairly recent.

File: nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/test/java/org/apache/nifi/provenance/TestPersistentProvenanceRepository.java (hunk 4 above)
Comment (markap14) [inline nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/test/java/org/apache/nifi/provenance/TestPersistentProvenanceRepository.java]: Whoops! Overlooked that. Was just a comment that I added in so that if i forgot to change the timeout back to 10,000 millis, the reviewer would catch it - glad it worked! :) Made it much longer so that i could set a breakpoint. Will change back to 10,000 millis and update the PR.
